### ISSUE TITLE
Unable to set oidcAuth clientSecret from Kubernetes Secret

### ISSUE THREAD

**Issue by kieranswhite** [gravitee-io/issues#8095]
## :collision: Describe the bug
When trying to set the oidcAuth options via Helm values I am unable to set the clientSecret to the value of a kubernetes secret. 

I'm attempting to use the secret configuration as described here: https://docs.gravitee.io/apim/3.x/apim_installguide_kubernetes.html#configuration 

## :sunrise_over_mountains: To Reproduce
Helm values configuration like the following:

```yaml
oidcAuth:
  enabled: true
  id: keycloak
  clientId: dev-gravitee-apm

  clientSecret: kubernetes://gravitee-apm/secrets/client-oidc-secret/client_oidc_secret

  tokenIntrospectionEndpoint: https://keycloak/auth/realms/myrealm/protocol/openid-connect/token/introspect
  tokenEndpoint: https://keycloak/auth/realms/myrealm/protocol/openid-connect/token
  authorizeEndpoint: https://keycloak/auth/realms/myrealm/protocol/openid-connect/auth
  userInfoEndpoint: https://keycloak/auth/realms/myrealm/protocol/openid-connect/userinfo
  userLogoutEndpoint: https://keycloak/auth/realms/myrealm/protocol/openid-connect/logout
  syncMappings: false #? What does this do?
  scopes:
    - openid
    - profile
    - email
  userMapping:
    id: preferred_username
    email: email
    lastname: family_name
    firstname: given_name
```

Kubernetes Secret Description:
```
Name:         client-oidc-secret
Namespace:    gravitee-apm
Labels:       <none>
Annotations:  <none>

Type:  Opaque

Data
====
client_oidc_secret:  36 bytes
```

Kubernetes Secret Value:
```
kubectl get secret client-oidc-secret --namespace gravitee-apm -o jsonpath='{.data}'
{"client_oidc_secret":"U3RBSUFRMDVOU0kwbnh5WllITGMyanliZGVRdzBiQllyTlZI"}%
```

Steps to reproduce the behaviour:
1. Go to the console login screen and attempt to login with Keycloak
2. The logs for gravitee API will show that the client secret is invalid
3. Additionally I can also see that value of the secret is incorrect from within my browser by changing the css to remove the `type="password", removing the *** replacement`

CSS Element: 
  ```
  <input _ngcontent-xaa-c205="" matinput="" name="configurationClientSecret" type="password" formcontrolname="clientSecret" required="" class="mat-input-element mat-form-field-autofill-control ng-tns-c141-9 ng-pristine cdk-text-field-autofill-monitored ng-valid ng-touched" ng-reflect-required="" ng-reflect-name="clientSecret" ng-reflect-type="password" id="mat-input-4" aria-required="true" aria-invalid="false">
  ```

Result of removing the `type="password"`
<img width="734" alt="image" src="https://user-images.githubusercontent.com/6289007/178528585-126fdcb9-c9bb-43fe-9ca5-1b500e7a3fb7.png">

## :rainbow: Expected behaviour
I would expect the kubernetes secret string to be replaced with the value of my secret

## Current behaviour
Currently I can't see this configuration setting being replaced with the value of the secret

## :movie_camera: Useful information
Screenshot, video, logs, other supporting material

## :computer: Desktop:

***Please complete the following information:***
 - Environment: 3.15.9 (Helm Version 3.1.51)
 - OS: Kubernetes 1.21
 - Browser Chrome

## :warning: Potential impacts

***Which other features may be impacted by this fix. This could be populated after fix***

***What are the impacted versions?***

## :link: Dependencies

Link a story or other related things...

**Comment by sfrevel** [comment]
Same problem here when using the uri for mongodb with the kubernetes: prefix like this
```
uri: kubernetes:/apim/secrets/mongodb-secret/uri
```
with k8s 1.22.6

**Comment by sfrevel** [comment]
Maybe this PropertyResolver is not initialized? https://github.com/gravitee-io/gravitee-node/blob/master/gravitee-node-kubernetes/src/main/java/io/gravitee/node/kubernetes/propertyresolver/KubernetesPropertyResolver.java

**Comment by sfrevel** [comment]
OK for me it looks like these properties are loaded with the wrong property resolver PropertySourcesPropertyResolver. From the logs i get
```
11:46:50.790 [graviteeio-node] DEBUG o.s.c.e.PropertySourcesPropertyResolver - Found key 'management.mongodb.uri' in PropertySource 'graviteeConfiguration' with value of type String
11:46:50.790 [graviteeio-node] DEBUG i.g.r.mongodb.common.MongoFactory - Read property management.mongodb.uri: kubernetes:/apim/secrets/mongodb-secret/uri
...
11:46:50.791 [graviteeio-node] WARN o.s.c.a.AnnotationConfigApplicationContext - Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'managementRepositoryConfiguration': Unsatisfied dependency expressed through field 'mappingMongoConverter'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'mappingMongoConverter' defined in io.gravitee.repository.mongodb.management.ManagementRepositoryConfiguration: Unsatisfied dependency expressed through method 'mappingMongoConverter' parameter 0; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'mongoDbFactory' defined in io.gravitee.repository.mongodb.management.ManagementRepositoryConfiguration: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.data.mongodb.MongoDatabaseFactory]: Circular reference involving containing bean 'managementRepositoryConfiguration' - consider declaring the factory method as static for independence from its containing instance. Factory method 'mongoDbFactory' threw exception; nested exception is java.lang.IllegalStateException: java.lang.IllegalArgumentException: The connection string is invalid. Connection strings must start with either 'mongodb://' or 'mongodb+srv://
```

**Comment by sfrevel** [comment]
hmmm, are these spaces in resources/META_INF/spring.factories really intended. I could not find any example with such spaces. All examples I found use a one liner or lines from 2 on are starting without spaces...

**Comment by kieranswhite** [comment]
FYI @sfrevel my current workaround has been to initialise ENV vars from secrets in the helm chart like so:

```
api:
  env:
      - name: gravitee_security_providers_2_clientSecret
        valueFrom:
          secretKeyRef:
            name: client-oidc-secret
            key: client_oidc_secret
            optional: false
```
See [here](https://docs.gravitee.io/pages/am/3.x/installation-guide/partial/how-to-configure.html#how_to_manage_arrays for ) for reference 

Then I've just set a dummy variable that will be overridden by the env var in the oidcAuth element:

```
oidcAuth:
  enabled: true
  id: keycloak
  clientSecret: "dummy"
```

**Comment by brasseld** [comment]
Hi guys,

Can you please enable the DEBUG logging level mode to see what's going wrong and provide the logs here ?

**Comment by sfrevel** [comment]
DEBUG was already on like in this post https://github.com/gravitee-io/issues/issues/8095#issuecomment-1227176531 - do you need some more log entries? I reduced it to the property where the problem seems to be...

**Comment by sfrevel** [comment]
> FYI @sfrevel my current workaround has been to initialise ENV vars from secrets in the helm chart like so:

Yes I this is also an option, but the RBAC is something which is  really smart, because the secrets are not in the environment variables

**Comment by sfrevel** [comment]
OK I also tested and debuged this in branch 3.18.x and found the the following log entry after turning on TRACE/DEBUG logging for org.springframework 
```
06:15:47.730 [graviteeio-node] DEBUG o.s.b.f.s.DefaultListableBeanFactory - Overriding bean definition for bean 'graviteeProperties' with a different definition: replacing [Root bean: class [io.gravitee.node.container.spring.env.PropertiesConfiguration]; scope=; abstract=false; lazyInit=null; autowireMode=3;
dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null;
factoryMethodName=graviteeProperties; initMethodName=null; destroyMethodName=(inferred); defined inio.gravitee.node.container.spring.env.PropertiesConfiguration] with [Root bean: class
[io.gravitee.rest.api.management.rest.spring.PropertiesConfiguration]; scope=; abstract=false; lazyInit=null;
autowireMode=3; dependencyCheck=0; autowireCandidate=true; primary=false; factoryBeanName=null;
factoryMethodName=graviteeProperties; initMethodName=null; destroyMethodName=(inferred); defined in
io.gravitee.rest.api.management.rest.spring.PropertiesConfiguration]
```
This led me to duplicated spring configurations which are overwriting the behavior of 
```
io.gravitee.node.container.spring.env
```
The two spring Configuration classes *RestManagementConfiguration*, *RestPortalConfiguration* from api's *StandaloneConfiguration* are overwriting beans and also imports their own *PropertiesConfiguration*

After removing all duplicated/overwriting bean definitions the right property handling mechanism is triggered and the property is loaded via the *KubernetesPropertyResolver* but still not loaded

An other problem came into place:
The gravitee.yml file was si configured to be loaded via 
```
GRAVITEE_OPTS="-Dgravitee.conf=/etc/gravitee-management/gravitee.yml"
```
in the file */gravitee-apim-rest-api-standalone-distribution/src/main/resources/bin/gravitee* 
but in the container this file is located under */opt/graviteeio-management-api/config/gravitee.yml*

After changing this to the opt-path the problem was gone and the properties are loaded from the kubernetes secret to the pod.

I could do a PR for this on the 3.18.x branch.

**Comment by sfrevel** [comment]
and here is the branch in my fork https://github.com/sfrevel/gravitee-api-management/tree/issue/8095-fix

**Comment by sfrevel** [comment]
So, is this fixed somewhere else, or did I miss something? I'm wondering why there is no response for the PR :(

**Comment by sfrevel** [comment]
Hi @brasseld 

As far as i remember I also had to fix the path, because the config file was not found in the given location

Regards

**Comment by gaetanmaisse** [comment]
@kieranswhite can you update to the latest `3.18` and share your feedback on this?

**Comment by gaetanmaisse** [comment]
Closing this issue as it should be fixed as of APIM 3.18.13 and there was no feedback for a while. Feels free to reopen it if needed.
